[MRG+2] switch to multinomial composition for mixture sampling

ljwolf opened the pull request:
<!--
Thanks for contributing a pull request! Please ensure you have taken a look at
the contribution guidelines: https://github.com/scikit-learn/scikit-learn/blob/master/CONTRIBUTING.md#Contributing-Pull-Requests
-->
#### Reference Issue

<!-- Example: Fixes #1234 -->

fixes #7701 
#### What does this implement/fix? Explain your changes.

This changes the way mixture models construct the composition of new samples. Specifically, any subclass deriving from`BaseMixture.sample` is affected. 

Instead of rounding the composition vector from `weights * n_samples`, this draws the composition vector from a multinomial distribution. Thus, samples return are guaranteed to have the number of observations requested by the user. However, the composition of the sample is now stochastic. 

In addition, this adds tests to ensure that `n_samples` are returned when `mixture.sample(n_samples)` is called. 

This may affect scaling for mixture models with a very large number of dimensions, since the multinational composition draw may be slow. But, this composition draw only occurs once during sampling. 
#### Any other comments?

None. Thanks for the great package!

<!--
Please be aware that we are a loose team of volunteers so patience is
necessary; assistance handling other issues is very welcome. We value
all user contributions, no matter how minor they are. If we are slow to
review, either the pull request needs some benchmarking, tinkering,
convincing, etc. or more likely the reviewers are simply busy. In either
case, we ask for your understanding during the review process.
For more information, see our FAQ on this topic:
http://scikit-learn.org/dev/faq.html#why-is-my-pull-request-not-getting-any-attention.

Thanks for contributing!
-->

ljwolf committed:
switch to multinomial composition for mixture sampling

tguillemot commented:
Can you add a test with the example you've put in #7701 on the function `test_sample` in `mixture/test_gaussian_mixture.py` ?
@ljwolf Thanks.

lesteve commented:
> This may affect scaling for mixture models with a very large number of dimensions, since the multinational composition draw may be slow.

This may be a problem, not sure.

This reminds me the issue we had in StratifiedShuffleSplit https://github.com/scikit-learn/scikit-learn/pull/6472 where drawing samples for each group did not add up to the total number of samples. Although I didn't really understand the details I believe @amueller used some kind of approximation to avoid randomly sampling.

tguillemot commented:
>This may affect scaling for mixture models with a very large number of dimensions, since the multinational composition draw may be slow.

It's not a problem. The function `sample` is not use during the fitting process. 
Moreover, as `weights` is linked to the number of component and `GaussianMixture` is not designed to deal with a lot of component (less than 1e6), the time is really OK.

On my computer, for a weights of shape = (1000000, ) :
```python
%timeit np.random.multinomial(100000000, weights).astype(int)
10 loops, best of 3: 135 ms per loop
```

jnothman commented on sklearn/mixture/base.py:
```diff
@@ -385,7 +385,7 @@ def sample(self, n_samples=1):
         _, n_features = self.means_.shape
         rng = check_random_state(self.random_state)
-        n_samples_comp = np.round(self.weights_ * n_samples).astype(int)
+        n_samples_comp = rng.multinomial(n_samples, self.weights_).astype(int)
```
the astype should not be necessary.

jnothman reviewed (dismissed):
I think this is an acceptable use of `rng.multinomial`. Please add a test.

amueller commented:
Can you please add a non-regression test? Otherwise looks good. Thanks for the fix!

ljwolf commented:
Yes, will do.

ljwolf committed:
add shape assertions to test

ljwolf commented:
I don't think I have the correct permissions to restart the travis build, but [it should be passing](https://travis-ci.org/ljwolf/scikit-learn). Can a maintainer trigger a rebuild?

jnothman commented on sklearn/mixture/tests/test_gaussian_mixture.py:
```diff
@@ -956,6 +957,13 @@ def test_sample():
 
+        # Check that sizes that are drawn match what is requested
+        assert_equal(X_s.shape, (n_samples, n_components))
+        for sample_size in [4, 101, 1004, 5051]:
```
Is there a particular reason to try with a large number? Would `for sample_size in range(50)` suffice?

ljwolf replied:
I think either would be sufficient. Earlier in the test, the sample size is 20,000. it should be fine to do `range(1,k)`, too. Should I use that instead?

jnothman commented:
As long as the test runs quite quickly, this LGTM.

jnothman commented:
Please update what's new

ljwolf commented:
I just swapped to the `range(1,50)`construct and added the fact that the test was added to the original statement. That Travis report looks pretty strange... tons of errors that look like build errors?

jnothman commented:
Could you please try updating master and rebasing on it?

ljwolf commented:
Alright, I've rebased to master & the tests are off to the races.

tguillemot commented:
LGTM

lesteve commented on sklearn/mixture/tests/test_gaussian_mixture.py:
```diff
@@ -956,6 +957,13 @@ def test_sample():
 
+        # Check that sizes that are drawn match what is requested
+        assert_equal(X_s.shape, (n_samples, n_components))
+        for sample_size in range(1, 50):
```
This text does not fail on master so it looks like you are not testing the edge case you discovered.

lesteve reviewed (approved):
I used n_components=3 for the test to test the actual regression seen in #7701.

As part of this it seems that n_features and n_components were swapped in a few places. @tguillemot can you quickly check whether what I changed makes sense.

lesteve commented:
> I used n_components=3 for the test to test the actual regression seen in #7701.

I forgot to say it explicitly: I used the recent github features that allows people with enough admin rights to push into PR branches. @ljwolf I hope you don't mind :-).

lesteve committed:
Use n_components=3 to test actual regression

n_components and n_features were equal and one was used for the other in
some places.

tguillemot commented:
@lesteve Sorry for these mistakes. It's `n_components` indeed.

amueller commented:
LGTM

ljwolf commented:
@lesteve nope, that's what they're there for! :shipit:

lesteve commented:
AppVeyor is taking quite some time, this one should be merged if it comes back green.

lesteve commented:
Merged, thanks a lot!

amueller commented:
@lesteve for future reference, please use the squash and merge feature, that makes cherry-picking much simpler.

lesteve commented:
Oops sorry for that, I always try to remember to do squash and merge but it looks like I missed this one.

lesteve commented:
Looks like you can only allow squash and merge if you want to:
https://github.com/blog/2141-squash-your-commits

The settings are available from: https://github.com/scikit-learn/scikit-learn/settings

Should we do that?

amueller commented:
@lesteve yes. I enabled that. Hm though sometimes we have multiple authors? Well, let's see when the first person complains. We can always go back. Thanks for finding that.

jnothman commented:
Whoops, we'd forgotten to tag this for 0.18.1. Done now.

amueller commented:
hm sorry didn't make it into 0.18.1. My bad :(
